src/backend/cdb/cdbpath.c

```diff
@@ -243,6 +242,47 @@ cdbpath_create_motion_path(PlannerInfo *root,
 			subpath->locus = locus;
 			return subpath;
 		}
+
+		if (CdbPathLocus_IsSingleQE(subpath->locus))
+		{
+			/*
+			 * For correct changing locus from SingleQE to Outer, we should add
+			 * Motion and Materialize nodes. If these nodes are present, just
+			 * change locus to Outer. Also there may be projection path node,
+			 * just change locus at the such node, too. For other cases Motion
+			 * and Materialize nodes will added at the end of this function.
+			 */
+
+			/*
+			 * Just one Motion node. Change locus to Outer and add Materialize.
+			 */
+			if (IsA(subpath, CdbMotionPath))
+			{
+				subpath->locus = locus;
+				return (Path *) create_material_path(root, subpath->parent,
+													 subpath);
+			}
+			/*
+			 * Projection Path and Motion. Change locus to Outer and add
+			 * Materialize between Projection and Motion.
+			 */
+			else if (IsA(subpath, ProjectionPath))
+			{
+				ProjectionPath* projection_path = (ProjectionPath *) subpath;
+
+				if (projection_path->subpath &&
+					IsA(projection_path->subpath, CdbMotionPath))
```

Comment:
Perhaps use `PointerIsValid(...)` or `... != NULL` to avoid confusing the field with a boolean condition?
```suggestion
				if (PointerIsValid(projection_path->subpath) &&
					IsA(projection_path->subpath, CdbMotionPath))
```